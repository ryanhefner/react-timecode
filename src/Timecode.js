import React, { forwardRef, useMemo } from 'react'
import cleanReactProps from 'clean-react-props'
import { formatTimecode } from './utils'

/**
 * Timecode - React Component
 *
 * @param {Object} props
 * @param {String} [props.as='span'] - HTML element to render
 * @param {String} [props.format='H:?m:ss'] - Timecode format
 * @param {String} [props.postfix=''] - Postfix to append to timecode
 * @param {String} [props.prefix=''] - Prefix to prepend to timecode
 * @param {Number} [props.time=0] - Time in milliseconds
 * @param {Object} [rest] - Additional props to pass to HTML element
 * @param {React.Ref} [ref] - React ref
 * @returns {React.ReactElement}
 */
export const Timecode = forwardRef(({ as: asProp = 'span', format = 'H:?m:ss', postfix = '', prefix = '', time = 0, ...rest }, ref) => {
  const timecode = useMemo(() => formatTimecode({ format, time }), [format, time])

  const Component = asProp

  return (
    <Component {...cleanReactProps(rest)} ref={ref}>
      {`${prefix}${timecode}${postfix}`}
    </Component>
  )
})
