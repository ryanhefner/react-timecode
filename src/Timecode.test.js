import React from 'react'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom'
import { Timecode } from './Timecode'

describe('<Timecode />', () => {
  test('render markup - default component', async () => {
    const { container, getByText } = render(<Timecode />)

    expect(getByText('0:00')).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  test('render markup - custom component', async () => {
    const { container } = render(<Timecode as="p" />)

    expect(container.firstChild.nodeName).toBe('P')
  })

  test('render format - H:?m:ss (default)', async () => {
    const { getByText } = render(<Timecode />)

    expect(getByText('0:00')).toBeInTheDocument()
  })

  test('render format - H:?mm:ss', async () => {
    const { getByText } = render(<Timecode format="H:?mm:ss" />)

    expect(getByText('00:00')).toBeInTheDocument()
  })

  test('render format - HH:mm:ss.SSS', async () => {
    const { getByText } = render(<Timecode format="HH:mm:ss.SSS" />)

    expect(getByText('00:00:00.000')).toBeInTheDocument()
  })

  test('render format - H:mm:ss.SSS', async () => {
    const { getByText } = render(<Timecode format="H:mm:ss.SSS" />)

    expect(getByText('0:00:00.000')).toBeInTheDocument()
  })

  test('render format - H:?mm:ss.SSS', async () => {
    const { getByText } = render(<Timecode format="H:?mm:ss.SSS" />)

    expect(getByText('00:00.000')).toBeInTheDocument()
  })

  test('render format - H:?m:ss.SSS', async () => {
    const { getByText } = render(<Timecode format="H:?m:ss.SSS" />)

    expect(getByText('0:00.000')).toBeInTheDocument()
  })

  test('render format - HH:mm:ss', async () => {
    const { getByText } = render(<Timecode format="HH:mm:ss" />)

    expect(getByText('00:00:00')).toBeInTheDocument()
  })

  test('render format - H:mm:ss', async () => {
    const { getByText } = render(<Timecode format="H:mm:ss" />)

    expect(getByText('0:00:00')).toBeInTheDocument()
  })

  test('render format - H:?mm:ss', async () => {
    const { getByText } = render(<Timecode format="H:?mm:ss" />)

    expect(getByText('00:00')).toBeInTheDocument()
  })

  test('render format - H:mm', async () => {
    const { getByText } = render(<Timecode format="H:mm" />)

    expect(getByText('0:00')).toBeInTheDocument()
  })

  test('render format - s.SSS', async () => {
    const { getByText } = render(<Timecode format="s.SSS" />)

    expect(getByText('0.000')).toBeInTheDocument()
  })

  test('render format - s.SS', async () => {
    const { getByText } = render(<Timecode format="s.SS" />)

    expect(getByText('0.00')).toBeInTheDocument()
  })

  test('render format w/ time - H:?m:ss (default)', async () => {
    const { getByText } = render(<Timecode time={3600000} />)

    expect(getByText('1:00:00')).toBeInTheDocument()
  })

  test('render format w/ time - H:?mm:ss.SSS', async () => {
    const { getByText } = render((
      <Timecode
        format="H:?mm:ss.SSS"
        time={3600000}
      />
    ))

    expect(getByText('1:00:00.000')).toBeInTheDocument()
  })

  test('render format w/ time - H:?m:ss.SSS', async () => {
    const { getByText } = render((
      <Timecode
        format="H:?m:ss.SSS"
        time={3600000}
      />
    ))

    expect(getByText('1:00:00.000')).toBeInTheDocument()
  })

  test('render format w/ time - H:?mm:ss', async () => {
    const { getByText } = render((
      <Timecode
        format="H:?mm:ss"
        time={3600000}
      />
    ))

    expect(getByText('1:00:00')).toBeInTheDocument()
  })

  test('render format w/ minutes - H:?m:ss (default)', async () => {
    const { getByText } = render(<Timecode time={5400000} />)

    expect(getByText('1:30:00')).toBeInTheDocument()
  })

  test('render format w/ seconds - H:?m:ss (default)', async () => {
    const { getByText } = render(<Timecode time={5430000} />)

    expect(getByText('1:30:30')).toBeInTheDocument()
  })

  test('render prefix', async () => {
    const { getByText } = render(<Timecode prefix="-" />)

    expect(getByText('-0:00')).toBeInTheDocument()
  })

  test('render postfix', async () => {
    const { getByText } = render(<Timecode postfix="/left" />)

    expect(getByText('0:00/left')).toBeInTheDocument()
  })
})
