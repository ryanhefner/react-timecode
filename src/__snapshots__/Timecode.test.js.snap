// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Timecode /> render markup - default component 1`] = `
<div>
  <span>
    0:00
  </span>
</div>
`;
